drivers/clock_control/clock_control_ast10x0.c

```diff
@@ -0,0 +1,167 @@
+#define   I3C_CLK_DIV_SEL		GENMASK(30, 28)
+#define     I3C_CLK_DIV_REG_TO_VAL(x)	((x == 0) ? 2 : (x + 1))
+#define   PCLK_DIV_SEL			GENMASK(11, 8)
+#define     PCLK_DIV_REG_TO_VAL(x)	((x + 1) << 1)
+#define CLK_SELECTION_REG5		0x314
+#define   HCLK_DIV_SEL			GENMASK(30, 28)
+#define     HCLK_DIV_REG_TO_VAL(x)	((x == 0) ? 2 : x + 1)
+
+struct clock_aspeed_config {
+	const struct device *syscon;
+};
+
+#define DEV_CFG(dev) ((const struct clock_aspeed_config *const)(dev)->config)
+
+static int aspeed_clock_control_on(const struct device *dev, clock_control_subsys_t sub_system)
+{
+	const struct device *syscon = DEV_CFG(dev)->syscon;
+	uint32_t clk_gate = (uint32_t)sub_system;
+	uint32_t addr = CLK_STOP_CTRL0_CLEAR;
+
+	/* there is no on/off control for group2 clocks */
+	if (clk_gate >= ASPEED_CLK_GRP_2_OFFSET) {
+		return 0;
+	}
+
+	if (clk_gate >= ASPEED_CLK_GRP_1_OFFSET) {
+		clk_gate -= ASPEED_CLK_GRP_1_OFFSET;
+		addr = CLK_STOP_CTRL1_CLEAR;
+	}
+
+	syscon_write_reg(syscon, addr, BIT(clk_gate));
+
+	return 0;
+}
+
+static int aspeed_clock_control_off(const struct device *dev, clock_control_subsys_t sub_system)
+{
+	const struct device *syscon = DEV_CFG(dev)->syscon;
+	uint32_t clk_gate = (uint32_t)sub_system;
+	uint32_t addr = CLK_STOP_CTRL0_SET;
+
+	/* there is no on/off control for group2 clocks */
+	if (clk_gate >= ASPEED_CLK_GRP_2_OFFSET) {
+		return 0;
+	}
+
+	if (clk_gate >= ASPEED_CLK_GRP_1_OFFSET) {
+		clk_gate -= ASPEED_CLK_GRP_1_OFFSET;
+		addr = CLK_STOP_CTRL1_SET;
+	}
+
+	syscon_write_reg(syscon, addr, BIT(clk_gate));
+
+	return 0;
+}
+
+static int aspeed_clock_control_get_rate(const struct device *dev,
+					 clock_control_subsys_t sub_system, uint32_t *rate)
+{
+	const struct device *syscon = DEV_CFG(dev)->syscon;
+	uint32_t clk_id = (uint32_t)sub_system;
+	uint32_t reg, src, clk_div;
+
+	switch (clk_id) {
+	case ASPEED_CLK_I3C0:
+	case ASPEED_CLK_I3C1:
+	case ASPEED_CLK_I3C2:
+	case ASPEED_CLK_I3C3:
+		syscon_read_reg(syscon, CLK_SELECTION_REG4, &reg);
+		if (FIELD_GET(I3C_CLK_SRC_SEL, reg) == I3C_CLK_SRC_HPLL) {
+			src = HPLL_FREQ;
+		} else {
+			src = 480000000;
+		}
+		clk_div = I3C_CLK_DIV_REG_TO_VAL(FIELD_GET(I3C_CLK_DIV_SEL, reg));
+		*rate = src / clk_div;
+		break;
+	case ASPEED_CLK_HCLK:
+		src = HPLL_FREQ;
+		syscon_read_reg(syscon, CLK_SELECTION_REG5, &reg);
+		clk_div = HCLK_DIV_REG_TO_VAL(FIELD_GET(HCLK_DIV_SEL, reg));
+		*rate = src / clk_div;
+		break;
+	case ASPEED_CLK_PCLK:
+		src = HPLL_FREQ;
+		syscon_read_reg(syscon, CLK_SELECTION_REG4, &reg);
+		clk_div = PCLK_DIV_REG_TO_VAL(FIELD_GET(PCLK_DIV_SEL, reg));
+		*rate = src / clk_div;
+		break;
+	case ASPEED_CLK_UART1:
+	case ASPEED_CLK_UART2:
+	case ASPEED_CLK_UART3:
+	case ASPEED_CLK_UART4:
+	case ASPEED_CLK_UART5:
+	case ASPEED_CLK_UART6:
+	case ASPEED_CLK_UART7:
+	case ASPEED_CLK_UART8:
+	case ASPEED_CLK_UART9:
+	case ASPEED_CLK_UART10:
+	case ASPEED_CLK_UART11:
+	case ASPEED_CLK_UART12:
+	case ASPEED_CLK_UART13:
+		*rate = 24000000 / 13;
```

Comment:
> So looks good.
> 
> Only question is about the fixed frequencies in here. I think a comment or two would be good. For example how or why is it that HPLL_FREQ is 1000Mhz? Why is I2C if not coming from HPLL, 480Mhz, and why UART is fixed 24Mhz/13.

These fixed frequencies are based on the hardware design.  

- HPLL is the clock source for CPU and BUS usage so it is fixed and unable to changed.

- The I3C clock is not used as the SCL directly.  Instead, it will be further divided into the other slow clocks in I3C IP according to the configuration of the I3C registers.

- The UART clock is not also used as the BAUD.  It will be further divided by the UART IP.